src/qibo/derivative.py

```diff
@@ -0,0 +1,125 @@
+                return ham
+
+            # defining a dummy circuit
+            def circuit(nqubits = 1):
+                c = Circuit(nqubits = 1)
+                c.add(gates.RY(q = 0, theta = 0))
+                c.add(gates.RX(q = 0, theta = 0))
+                c.add(gates.M(0))
+                return c
+
+            # using GradientTape to benchmark
+            def gradient_tape(params):
+                params = tf.Variable(params)
+
+                with tf.GradientTape() as tape:
+                    c = circuit(nqubits = 1)
+                    c.set_parameters(params)
+                    h = hamiltonian()
+                    expected_value = h.expectation(c.execute().state())
+
+                grads = tape.gradient(expected_value, [params])
+                return grads
+
+            # initializing the circuit
+            c = circuit(nqubits = 1)
+
+            # some parameters
+            test_params = np.random.randn(2)
+            c.set_parameters(test_params)
+
+            # a test hamiltonian
+            test_hamiltonian = hamiltonian()
+
+            # running the psr with respect to the two parameters
+            grad_0 = parameter_shift(circuit = c, observable = test_hamiltonian, parameter_index = 0, gate_eigenv = 0.5)
+            grad_1 = parameter_shift(circuit = c, observable = test_hamiltonian, parameter_index = 1, gate_eigenv = 0.5)
+
+            # evaluating the gradients using tensorflow
+            tf_grads = gradient_tape(test_params)
+
+            print('Test gradient with respect params[0] with PSR: ', grad_0.numpy())
+            print('Test gradient with respect params[0] with tf:  ', tf_grads[0][0].numpy())
+            print('Test gradient with respect params[1] with PSR: ', grad_1.numpy())
+            print('Test gradient with respect params[1] with tf:  ', tf_grads[0][1].numpy())
+
+    """
+
+    if parameter_index > len(circuit.get_parameters()):
+        raise ValueError(
+            """This index doesn't exist! Use one of the circuit's parameters."""
+        )
+
+    # defining the shift according to the psr
+    s = np.pi / (4 * gate_eigenv)
+
+    # saving original parameters and making a copy
+    original = np.asarray(circuit.get_parameters()).copy()
+    shifted = original.copy()
+
+    # forward shift and evaluation
+    shifted[parameter_index] += s
+    circuit.set_parameters(shifted)
+
+    if initial_state is None:
+        forward = observable.expectation(circuit.execute().state())
+    else:
+        forward = observable.expectation(
+            circuit.execute(initial_state=initial_state).state()
+        )
+
+    # backward shift and evaluation
+    shifted[parameter_index] -= 2 * s
+    circuit.set_parameters(shifted)
+
+    if initial_state is None:
+        backward = observable.expectation(circuit.execute().state())
+    else:
+        backward = observable.expectation(
+            circuit.execute(initial_state=initial_state).state()
+        )
+
```

Comment:
Diego is right, this `if`/`else` is not needed. You can pass the `initial_state` in the circuit execution even if it is `None`. If `None` the circuit will automatically use |00...0> as you want:
```suggestion
    backward = observable.expectation(backend.execute_circuit(circuit, initial_state=initial_state).state())
```
Same for the `forward` propagation above.

Reply:
Done, thank you!